SwiftCompilerSources for ownership, borrowing, and liveness

atrick opened the pull request:
Add BorrowUtils: OSSA infrastructure for borrow scopes.
    Key APIs necessary for using OSSA.
    - BorrowingInstruction
    - BeginBorrowValue
    - scopeEndingOperands
    - BorrowIntroducers
    - EnclosingValues
    - innerAdjacentPhis

Add OwnershipLiveness.swift. Core OSSA APIs.
    This is what you need to correctly analyze OSSA.
    - computeLinearLiveness
    - computeInteriorLiveness
    - InteriorUseVisitor
    - OwnershipUseVisitor
    - LivenessBoundary

This incorporates review feedback from
https://github.com/apple/swift/pull/70460

atrick committed:
ForwardingUtils comment formatting.

Format comments so they are readable on github and 80-col tools.

atrick committed:
ForwardingUtils.swift cleanup. Make the API easier to extend.

atrick committed:
Add BorrowUtils: OSSA infrastructure for borrow scopes.

Key APIs necessary for using OSSA.

- BorrowingInstruction

- BeginBorrowValue

- scopeEndingOperands

- BorrowIntroducers

- EnclosingValues

- innerAdjacentPhis

These need to be complete to be correct.

atrick committed:
Add OwnershipLiveness.swift. Core OSSA APIs.

This is what you need to correctly analyze OSSA.

- computeLinearLiveness

- computeInteriorLiveness

- InteriorUseVisitor

- OwnershipUseVisitor

- LivenessBoundary

Along with BorrowUtils.swift, all of our OSSA transformations are
built on top of these fundamentals. With these APIs, we can build
anything OSSA-related in SwiftCompilerSources. These utilities are
immediately needed for borrowed arguments and lifetime dependence. In
the near future, we can also use them to complete OSSA lifetimes and
*correctly* fixup OSSA after transformation without introducing lots
of copies and creating lots of incorrect corner cases.

atrick committed:
Add OperandOwnership.endsLifetime

And rename `uses.lifetimeEndingUses` to `uses.endingLifetime`

atrick committed:
BorrowIntroducer unit tests

atrick committed:
Add Stack.Marker and Stack.Segment utilities.

Also fix the memory binding within stack access.

atrick committed:
Add SingleInlineArray utility.

atrick committed:
Ownership liveness in Swift unit tests.

atrick committed:
Add Swift SIL instructions needed by AddressUseVisitor

atrick committed:
SwiftCompilerSources: add UseList ignoreTypeDependence

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test

atrick commented:
This PR is blocked by a likely recent LoadableByAddress bug that happens to be exposed.

atrick commented:
@swift-ci test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+  /// result and implement this as a fatalError.
+  mutating func visitPointerEscape(use: Operand) -> WalkResult
+
+  /// Handle begin_borrow, load_borrow, store_borrow, begin_apply.
```
I'm completely confused by this comment.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+  }
+}
+
+/// Enumerate all special cases of ownership uses in a visitor
```
It's not clear to me what `OwnershipUseVisitor` does or should do. Also this comment is not really useful.
On the one hand it appears to wrap a visitor around the `OperandOwnership` enum. On the other hand it looks like it's also doing some kind of def-use walk.
The confusion already starts with that I'm not sure what the terms "outer" and "inner" scope/liferange/value actually refer to.

atrick replied:
`OwnershipUseVisitor` is the one place that define the effects of operand ownership. We need this so we don't end up with ad-hoc utilities in OptUtils like `InstructionRange.insert(borrowScopeOf)` which is out of place, incomplete, and likely misused.

eeckstein replied:
Still, I don't know what an `OwnershipUseVisitor` does and what I can use it for.
Small and simple utilities like `InstructionRange.insert(borrowScopeOf)` are much easier to use because it's immediately clear what they do.
We really need to keep those new utilities understandable and easy to use. Part of it is good API names and good comments. If the purpose of a utility cannot be expressed in a few sentences in a comment its design is probably too complex.

eeckstein reviewed (commented):
The APIs and utilities in OwnershipLiveness still need quite a lot of polishing, clarification and better documentation.

Note that I also added a few comments in the old PR https://github.com/apple/swift/pull/70460

atrick committed:
AddressUseVisitor

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift:
```diff
@@ -0,0 +1,655 @@
+/// If `value` is a reborrow, then this either returns a dominating
+/// enclosing value or an outer adjacent phi.
+///
+/// Example:                                       // enclosing value:
```
That logic is stated in the simplest form in `EnclosingValues.gather`. It is called recursively with its own state for tracking reborrows.

Those top-level utility functions don't do anything but create a Stack state that's needed for analysis.

atrick replied:
Note that `EnclosingValues` deals with reborrows, and `BorrowIntroducers` deals with guaranteed forwarding phis. It's a similar problem, so they both use `mapToPhi`. But it's also useful to know which you're handling, and splitting the utilities made it much easier for me to reason about the state of recursion.

eeckstein replied:
> That logic is stated in the simplest form in EnclosingValues.gather.

Right. But why is `EnclosingValues.gather(forReborrow:)` needed? Aren't enclosing values of a reborrow the same as its borrow introducers? So, shouldn't `BorrowIntroducers.gather` find the enclosing values?

atrick replied:
EnclosingValues deals with reborrows, and BorrowIntroducers deals with guaranteed forwarding phis. It's a similar problem, so they both use `mapToPhi`. But it's also useful to know which you're handling, and splitting the utilities made it much easier for me to reason about the state of recursion.

atrick committed:
OwnershipUtils review feedback: Comments and examples.

atrick committed:
OwnershipUtils review feedback: code formatting.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+
+/// Indicate whether OwnershipUseVisitor is visiting a use of the
+/// outer OSSA lifetime or within an inner borrow scope (reborrow).
+enum IsInnerLifetime {
```
I originally wrote this as you suggested but found an enum be less bug prone. There's no obvious true/false direction for the boolean in this case, and it's easy to accidentally reverse the meaning, even with argument labels.

At this point, I'm mostly relying on it as a good place for comments and to make the code more readable.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+  assert(visitor.unenclosedPhis.isEmpty, "missing adjacent phis")
+  return range
+}
+
```
Forwarding is a simple, well-defined concept that is independent from any other OSSA concepts. The analysis of borrow scopes relies on forwarding. But it's often sufficient to just check ForwardingInstruction, which is much simpler than conforming to a walker utility. Sometimes LifetimeDependence also needs to know about forwarding, but it is also a separate concept.

This is an example of why composition is better than inheritance. Each utility does exactly one thing, which is very simple and well-defined but can be used in many different contexts.

The biggest problem with our previous utilities is that the don't cleanly model well-defined concepts. We merge utilities to reduce lines of code and in the process generate overwhelming complexity. A single utility that means different things in different contexts is the worst kind of complexity. It's like using a type for something that is not that type just because it shares a representation.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+  let context: FunctionPassContext
+  var _context: Context { context }
+
+  let function: Function
```
I don't need a `function` property, but I like all stateful utilities to have a `function` property. Figuring out how to access the current function in each line of code is one of the most unnecessarily ridiculous things in LLVM. It makes the code much harder to read, work with, and refactor. In SwiftCompilerSources, I'm confused by why we can't ask the context for the current function. But short of that, utilities should provide it.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+///
+/// TODO: Verify that pointerEscape is only called for live ranges in which
+/// `findPointerEscape()` returns true.
+protocol AddressUseVisitor {
```
That Walker convention is good but I don't want to rename AddressUseVisitor yet. Its purpose of is to visit all of the address uses, not just the leaf uses, partitioning them by the kind of use. This is the most general possible address use visitor, so there's no other sensible name.

Note that I originally wrote `AddressUseVisitor` as an extension of `AddressDefUseWalker`, even though it's obviously backward, just to keep up the appearance of reuse. It was complex and bug-prone, and the results were invalid.

For now, I need to separate this new infrastructure from the code that affects the rest of the compiler. So redesigning `AddressDefUseWalker` should be a separate step. In fact, the separate list of opcodes is necessary until we want to change the behavior of existing passes.

Literally the only thing you want to reuse here is the list of address projection opcodes. The simple, obvious way to do that is with an AddressProjection abstraction. That would solve any short term problem. In general, forcing inheritance to reuse logic leads to incomprehensible, buggy code.

Longer term, the two utilities should be integrated. There are multiple problems with the current design and it's not possible to write a simple address use visitor on top of it. One thing I'll point out is that `AddressDefUseWalker.leafUse` is just outright wrong. It's called on just any unrecognizable use, and that's not suitable for utilities that are designed to be complete.

`AddressDefUseWalker` should be built on top of `AddressUseVisitor`. It should simply override `projection` to update the path and it should be responsible for the def-use walk itself, rather than just visiting the use. Then AddressUseVisitor won't have to do any walk and everything will make sense and work naturally.

eeckstein replied:
> The simple, obvious way to do that is with an AddressProjection abstraction.

AddressDefUseWalker _is_ the address projection abstraction. Its implementation mainly consists of a switch over all address projection instructions and "translating" that to a projection path.

atrick replied:
I'll use `AddressLifetimeDefUseWalker` for this. Later, I think the other one should be renamed`AddressPathDefUseWalker` or `AddressProjectionDefUseWalker`. We should still add an `AddressProjection` protocol.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,977 @@
+         is ProjectBlockStorageInst, is TailAddrInst, is StoreBorrowInst,
+         is UncheckedAddrCastInst, is MarkUninitializedInst, is DropDeinitInst,
+         is TuplePackElementAddrInst, is CopyableToMoveOnlyWrapperAddrInst,
+         is MoveOnlyWrapperToCopyableAddrInst:
```
As mentioned above, we should have one list of address projections, and it definitely should not be buried inside a walker! But for now, the two implementations are intentionally not in sync with each other. SIL verification should come first.

atrick commented:
@swift-ci test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift:
```diff
@@ -0,0 +1,780 @@
+      return
+    }
+    // avoid duplicates in the enclosingValues set.
+    var pushedEnclosingValues = ValueSet(context)
```
You cannot use `ValueSet` in a recursive function. It will run out of bit positions (for the set) pretty quickly.
Can you define `pushedEnclosingValues` as member property of `EnclosingValues`?

atrick replied:
Thanks, i meant to convert that ValueSet into a cache. That's done now.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipLiveness.swift:
```diff
@@ -0,0 +1,1057 @@
+    }
+  }
+
+  /// Visit only those uses of an value within an inner borrow scope
```
`a value`

atrick committed:
Add a reborrow cache to EnclosingValues.

Along with motivating unit tests.

atrick committed:
Reorder code in OwnershipUtils.swift.

To occur in the order that people will likely want to read.

atrick committed:
Rename AddressUseVisitor to AddressLifetimeDefUseWalker.

atrick committed:
Update SIL-Utilities.md for new ownership utils.

atrick commented on SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift:
```diff
@@ -0,0 +1,780 @@
+//     %6 = begin_borrow %0       %6                   %0
+//     %7 = struct (%6)           %6                   %6
+//     br bb3(%6, %7)             
+//   bb3(%reborrow: @guaranteed,  %reborrow            %0
```
That would be great. For some reason we still print `@guaranteed`. I'll change the comments to `@reborrow` to anticipate fixing this.

atrick committed:
Update OwnershipLiveness comments for code review.

atrick commented:
@swift-ci test

eeckstein reviewed (approved):
lgtm!

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift:
```diff
@@ -0,0 +1,832 @@
+    case let arg as Argument where arg.isReborrow:
+      self = .reborrow(Phi(arg)!)
+    default:
+      if value.definingInstruction is BeginApplyInst {
```
nit, in case you are pushing a new version anyway: you can write
```
  case let mvir as MultipleValueInstructionResult where mvir.parentInstruction is BeginApplyInst:
```
Then you can also define the case payload as `case beginApply(MultipleValueInstructionResult)` which avoids the `!` when getting its parent instruction.

atrick commented:
@swift-ci smoke test

atrick commented:
@swift-ci smoke test

atrick commented:
@swift-ci smoke test linux

atrick commented:
@swift-ci smoke test windows

atrick commented:
@swift-ci smoke test linux

atrick commented:
@swift-ci smoke test windows

atrick committed:
Disable all new tests on Windows.

We have language features that depends on swift_in_compiler. Any
platform that does not support this build feature does not really
support the language, so we must continue to disable tests.

atrick commented:
@swift-ci smoke test
